misc/experimental/promise/api/celix/impl/SharedPromiseState.h

```diff
@@ -0,0 +1,508 @@
+    std::unique_lock<std::mutex> lck{mutex};
+    if (done) {
+        throw celix::PromiseInvocationException("Cannot resolve Promise. Promise is already done");
+    }
+    new(&data) T{std::forward<T>(value)};
+    exp = nullptr;
+    complete(lck);
+}
+
+
+template<typename T>
+inline void celix::impl::SharedPromiseState<T>::resolve(const T& value) {
+    std::unique_lock<std::mutex> lck{mutex};
+    if (done) {
+        throw celix::PromiseInvocationException("Cannot resolve Promise. Promise is already done");
+    }
+    new(&data) T{value};
+    exp = nullptr;
+    complete(lck);
+}
+
+template<typename T>
+inline void celix::impl::SharedPromiseState<T>::fail(std::exception_ptr e) {
+    std::unique_lock<std::mutex> lck{mutex};
+    if (done) {
+        throw celix::PromiseInvocationException("Cannot fail Promise. Promise is already done");
+    }
+    exp = e;
+    complete(lck);
+}
+
+template<typename T>
+inline void celix::impl::SharedPromiseState<T>::fail(const std::exception& e) {
+    fail(std::make_exception_ptr(e));
+}
+
+template<typename T>
+inline void celix::impl::SharedPromiseState<T>::tryResolve(T&& value) {
+    std::unique_lock<std::mutex> lck{mutex};
+    if (!done) {
+        new(&data) T(std::forward<T>(value));
+        exp = nullptr;
+        complete(lck);
+    }
+}
+
+template<typename T>
+inline void celix::impl::SharedPromiseState<T>::tryFail(std::exception_ptr e) {
+    std::unique_lock<std::mutex> lck{mutex};
+    if (!done) {
+        exp = e;
+        complete(lck);
+    }
+}
+
+template<typename T>
+inline bool celix::impl::SharedPromiseState<T>::isDone() const {
+    std::lock_guard<std::mutex> lck{mutex};
+    return done;
+}
+
+template<typename T>
+inline bool celix::impl::SharedPromiseState<T>::isSuccessfullyResolved() const {
+    std::lock_guard<std::mutex> lck{mutex};
+    return done && !exp;
+}
+
+
+template<typename T>
+inline void celix::impl::SharedPromiseState<T>::waitForAndCheckData(std::unique_lock<std::mutex>& lck, bool expectValid) const {
+    if (!lck.owns_lock()) {
+        lck.lock();
+    }
+    cond.wait(lck, [this]{return done;});
+    if (expectValid && exp) {
+        throw celix::PromiseInvocationException{"Expected a succeeded promise, but promise failed"};
+    } else if(!expectValid && !exp && !dataMoved) {
+        throw celix::PromiseInvocationException{"Expected a failed promise, but promise succeeded"};
+    } else if (dataMoved) {
+        throw celix::PromiseInvocationException{"Invalid use of promise, data is moved and not available anymore!"};
+    }
+}
+
+template<typename T>
+inline const T& celix::impl::SharedPromiseState<T>::getValue() const {
+    std::unique_lock<std::mutex> lck{mutex};
+    waitForAndCheckData(lck, true);
+    const T* ptr = reinterpret_cast<const T*>(&data);
+    return *ptr;
+}
+
+template<typename T>
+inline tbb::task_arena celix::impl::SharedPromiseState<T>::getExecutor() const {
```

Comment:
Conceptually the Executor belongs to the PromiseFactory, and the Promise instances obtain the Executor from there.

Reply:
Correct, I will update this.

Reply:
executor is not injected from PromiseFactory. For now I think this is good enough.
In the near future we will look which thread pool library we want to use, if needed we will address this issue then.